File: logistics/utils/freight_agent_service.py
# Copyright (c) 2026, www.agilasoft.com and contributors
# For license information, please see license.txt

"""Validate Freight Agent links against applicable service types on operational documents."""

from __future__ import annotations

import frappe
from frappe import _

from logistics.utils.charge_service_type import (
	IMPLIED_SERVICE_TYPE_BY_DOCTYPE,
	implied_service_type_for_doctype,
	operational_booking_charge_service_type_label,
)
from logistics.utils.service_mode_flags import validate_service_mode_link

FREIGHT_AGENT_LINK_FIELDS = frozenset({"freight_agent", "freight_agent_sea"})

# Explicit field → service label when it cannot be inferred from doctype alone.
_FIELD_SERVICE_TYPE_LABELS: dict[str, str] = {
	"freight_agent_sea": "Sea",
	"air_default_sending_agent": "Air",
	"air_default_receiving_agent": "Air",
	"air_default_broker": "Air",
	"sea_default_sending_agent": "Sea",
	"sea_default_receiving_agent": "Sea",
	"customs_default_freight_agent": "Customs",
}

_SETTINGS_DOCTYPE_SERVICE: dict[str, str] = {
	"Air Freight Settings": "Air",
	"Sea Freight Settings": "Sea",
}

_DOCTYPE_HAS_FREIGHT_AGENT_CACHE: dict[str, bool] = {}


def _meta_has_freight_agent_reference(meta) -> bool:
	for df in meta.fields:
		if df.fieldtype == "Link" and df.options == "Freight Agent":
			return True
		if df.fieldtype == "Table":
			try:
				child_meta = frappe.get_meta(df.options)
			except Exception:
				continue
			for cdf in child_meta.fields:
				if cdf.fieldtype == "Link" and cdf.options == "Freight Agent":
					return True
	return False


def doctype_references_freight_agent(doctype: str) -> bool:
	if doctype in _DOCTYPE_HAS_FREIGHT_AGENT_CACHE:
		return _DOCTYPE_HAS_FREIGHT_AGENT_CACHE[doctype]
	try:
		meta = frappe.get_meta(doctype)
	except Exception:
		_DOCTYPE_HAS_FREIGHT_AGENT_CACHE[doctype] = False
		return False
	out = _meta_has_freight_agent_reference(meta)
	_DOCTYPE_HAS_FREIGHT_AGENT_CACHE[doctype] = out
	return out


def resolve_service_type_label_for_freight_agent(doc, fieldname: str, row=None) -> str | None:
	"""Return UI service label (Air, Sea, …) for validating a Freight Agent link."""
	ctx = row if row is not None else doc

	service_type = _read_value(ctx, "service_type")
	if service_type:
		return operational_booking_charge_service_type_label(service_type, default=service_type)

	if fieldname in _FIELD_SERVICE_TYPE_LABELS:
		return _FIELD_SERVICE_TYPE_LABELS[fieldname]

	main_service = _read_value(doc, "main_service")
	if main_service:
		return operational_booking_charge_service_type_label(main_service, default=main_service)

	doctype = getattr(doc, "doctype", None) or ""
	if doctype in _SETTINGS_DOCTYPE_SERVICE:
		return _SETTINGS_DOCTYPE_SERVICE[doctype]

	implied = implied_service_type_for_doctype(doctype)
	if implied:
		return implied

	if fieldname in FREIGHT_AGENT_LINK_FIELDS:
		return IMPLIED_SERVICE_TYPE_BY_DOCTYPE.get(doctype)

	return None


def validate_freight_agent_link(
	agent_name: str | None,
	service_type_label: str | None,
	*,
	context: str,
) -> None:
	validate_service_mode_link(
		"Freight Agent",
		agent_name,
		service_type_label,
		context=context,
	)


def validate_freight_agent_links_on_doc(doc, method=None) -> None:
	if getattr(frappe.flags, "in_install", None) or getattr(frappe.flags, "in_migrate", None):
		return
	if getattr(frappe.flags, "in_import", None):
		return
	if not doctype_references_freight_agent(doc.doctype):
		return

	prev = doc.get_doc_before_save() if not doc.is_new() else None
	_validate_parent_freight_agent_links(doc, prev)
	_validate_child_table_freight_agent_links(doc, prev)


def _validate_parent_freight_agent_links(doc, prev) -> None:
	meta = frappe.get_meta(doc.doctype)
	for df in meta.fields:
		if df.fieldtype != "Link" or df.options != "Freight Agent":
			continue
		val = doc.get(df.fieldname)
		if not val:
			continue
		old_val = prev.get(df.fieldname) if prev else None
		if prev and old_val == val:
			continue
		service_label = resolve_service_type_label_for_freight_agent(doc, df.fieldname)
		if not service_label:
			continue
		validate_freight_agent_link(
			val,
			service_label,
			context=_("{0} ({1})").format(df.label or df.fieldname, doc.doctype),
		)


def _validate_child_table_freight_agent_links(doc, prev) -> None:
	meta = frappe.get_meta(doc.doctype)
	prev_by_name: dict[tuple[str, str], object] = {}
	if prev:
		for df in meta.fields:
			if df.fieldtype != "Table":
				continue
			for row in prev.get(df.fieldname) or []:
				if row.name:
					prev_by_name[(df.fieldname, row.name)] = row

	for df in meta.fields:
		if df.fieldtype != "Table":
			continue
		child_meta = frappe.get_meta(df.options)
		link_fields = [
			cdf for cdf in child_meta.fields if cdf.fieldtype == "Link" and cdf.options == "Freight Agent"
		]
		if not link_fields:
			continue
		table_label = df.label or df.fieldname
		for row in doc.get(df.fieldname) or []:
			pr = prev_by_name.get((df.fieldname, row.name)) if row.name else None
			for lf in link_fields:
				val = row.get(lf.fieldname)
				if not val:
					continue
				old_val = pr.get(lf.fieldname) if pr else None
				if pr and old_val == val:
					continue
				service_label = resolve_service_type_label_for_freight_agent(doc, lf.fieldname, row=row)
				if not service_label:
					continue
				validate_freight_agent_link(
					val,
					service_label,
					context=_("{0} row {1} ({2})").format(table_label, row.idx, lf.label or lf.fieldname),
				)


def _read_value(doc, fieldname: str):
	if doc is None:
		return None
	if isinstance(doc, dict):
		return doc.get(fieldname)
	return getattr(doc, fieldname, None)

File: logistics/public/js/freight_agent_service.js
// Copyright (c) 2026, Agilasoft and contributors
// For license information, please see license.txt

/**
 * Filter Freight Agent link fields by applicable service types (Air, Sea, Transport, Customs, Warehousing).
 */
frappe.provide("logistics.freight_agent_service");

(function () {
	"use strict";

	const FIELD_SERVICE_MAP = {
		freight_agent_sea: "Sea",
		air_default_sending_agent: "Air",
		air_default_receiving_agent: "Air",
		air_default_broker: "Air",
		sea_default_sending_agent: "Sea",
		sea_default_receiving_agent: "Sea",
		customs_default_freight_agent: "Customs",
	};

	const SETTINGS_DOCTYPE_SERVICE = {
		"Air Freight Settings": "Air",
		"Sea Freight Settings": "Sea",
	};

	const DOCTYPE_IMPLIED_SERVICE = {
		"Air Booking": "Air",
		"Air Shipment": "Air",
		"Sea Booking": "Sea",
		"Sea Shipment": "Sea",
		"Transport Order": "Transport",
		"Transport Job": "Transport",
		Declaration: "Customs",
		"Declaration Order": "Customs",
		"Warehouse Job": "Warehousing",
		"VAS Order": "Warehousing",
		"Inbound Order": "Warehousing",
	};

	/** Services with no Freight Agent module checkbox (must not become filter fieldnames). */
	const MODULE_FLAGS_WITHOUT_SERVICE = new Set(["special project", "exhibits", "mice", "events"]);

	function canonicalChargeServiceType(st) {
		const raw = (st && String(st).trim()) || "";
		if (!raw) return "";
		const byTitle = {
			Air: "air",
			Sea: "sea",
			Transport: "transport",
			Custom: "custom",
			Customs: "custom",
			custom: "custom",
			customs: "custom",
			Warehousing: "warehousing",
			"Special Project": "special project",
			MICE: "exhibits",
			Exhibits: "exhibits",
			Events: "exhibits",
		};
		if (byTitle[raw] !== undefined) return byTitle[raw];
		const low = raw.toLowerCase();
		if (low === "customs") return "custom";
		if (low === "events" || low === "mice") return "exhibits";
		if (["air", "sea", "transport", "warehousing", "special project", "exhibits"].includes(low)) {
			return low;
		}
		return low;
	}

	/** Freight Agent checkbox field for a service type, or null when none applies. */
	function moduleFlagForServiceType(st) {
		const c = canonicalChargeServiceType(st);
		if (!c || MODULE_FLAGS_WITHOUT_SERVICE.has(c)) return null;
		if (c === "custom") return "customs";
		if (["air", "sea", "transport", "warehousing"].includes(c)) return c;
		return null;
	}

	/** Sales Quote main_service is a Select (Air/Sea/…); on linked bookings it is a Dynamic Link job ID. */
	function quoteStyleMainService(frm) {
		const df =
			(frm.get_docfield && frm.get_docfield("main_service")) ||
			frappe.meta.get_docfield(frm.doctype, "main_service");
		if (!df || df.fieldtype === "Dynamic Link" || !frm.doc.main_service) {
			return null;
		}
		return frm.doc.main_service;
	}

	function resolveServiceTypeLabel(frm, fieldname, row) {
		if (row && row.service_type) {
			return row.service_type;
		}
		if (FIELD_SERVICE_MAP[fieldname]) {
			return FIELD_SERVICE_MAP[fieldname];
		}
		const quoteMain = quoteStyleMainService(frm);
		if (quoteMain) {
			return quoteMain;
		}
		if (SETTINGS_DOCTYPE_SERVICE[frm.doctype]) {
			return SETTINGS_DOCTYPE_SERVICE[frm.doctype];
		}
		return DOCTYPE_IMPLIED_SERVICE[frm.doctype] || null;
	}

	function stripFreightAgentLinkFilters(doctype, fieldname, childFieldname) {
		const df = frappe.meta.get_docfield(doctype, fieldname, childFieldname);
		if (df && df.link_filters) {
			df.link_filters = null;
		}
	}

	function clearIncompatibleFreightAgent(frm, fieldname, serviceType, childFieldname, cdn) {
		const value = childFieldname
			? (locals[childFieldname] && locals[childFieldname][cdn] && locals[childFieldname][cdn][fieldname])
			: frm.doc[fieldname];
		if (!value) return;
		const flag = moduleFlagForServiceType(serviceType);
		if (!flag) return;
		frappe.db.get_value("Freight Agent", value, flag, (r) => {
			if (r && r[flag]) return;
			if (childFieldname && cdn) {
				frappe.model.set_value(childFieldname, cdn, fieldname, "");
			} else {
				frm.set_value(fieldname, "");
			}
		});
	}

	function setupParentFreightAgentQueries(frm) {
		(frm.meta.fields || []).forEach((df) => {
			if (df.fieldtype !== "Link" || df.options !== "Freight Agent") return;
			stripFreightAgentLinkFilters(frm.doctype, df.fieldname);
			frm.set_query(df.fieldname, function () {
				stripFreightAgentLinkFilters(frm.doctype, df.fieldname);
				const filters = { is_active: 1 };
				const serviceType = resolveServiceTypeLabel(frm, df.fieldname);
				const flag = moduleFlagForServiceType(serviceType);
				if (flag) {
					filters[flag] = 1;
				}
				return { filters };
			});
		});
	}

	function setupChildFreightAgentQueries(frm) {
		(frm.meta.fields || []).forEach((tableDf) => {
			if (tableDf.fieldtype !== "Table") return;
			const childMeta = frappe.get_meta(tableDf.options);
			if (!childMeta) return;
			(childMeta.fields || []).forEach((cdf) => {
				if (cdf.fieldtype !== "Link" || cdf.options !== "Freight Agent") return;
				stripFreightAgentLinkFilters(tableDf.options, cdf.fieldname, tableDf.fieldname);
				frm.set_query(cdf.fieldname, tableDf.fieldname, function (doc, cdt, cdn) {
					stripFreightAgentLinkFilters(tableDf.options, cdf.fieldname, tableDf.fieldname);
					const row = (locals[cdt] && locals[cdt][cdn]) || doc;
					const filters = { is_active: 1 };
					const serviceType = resolveServiceTypeLabel(frm, cdf.fieldname, row);
					const flag = moduleFlagForServiceType(serviceType);
					if (flag) {
						filters[flag] = 1;
					}
					return { filters };
				});
			});
		});
	}

	logistics.freight_agent_service.moduleFlagForServiceType = moduleFlagForServiceType;
	logistics.freight_agent_service.resolveServiceTypeLabel = resolveServiceTypeLabel;
	logistics.freight_agent_service.clearIncompatibleFreightAgent = clearIncompatibleFreightAgent;

	logistics.freight_agent_service.setupForm = function (frm) {
		if (!frm || !frm.meta) return;
		setupParentFreightAgentQueries(frm);
		setupChildFreightAgentQueries(frm);
	};

	frappe.ui.form.on("*", {
		refresh(frm) {
			logistics.freight_agent_service.setupForm(frm);
		},
		main_service(frm) {
			["freight_agent", "freight_agent_sea"].forEach((fieldname) => {
				if (!frm.fields_dict[fieldname]) return;
				clearIncompatibleFreightAgent(frm, fieldname, resolveServiceTypeLabel(frm, fieldname));
			});
		},
	});
})();
